hack/jenkins/e2e-runner.sh

```diff
@@ -21,8 +21,54 @@ set -o nounset
 set -o pipefail
 set -o xtrace
 
+# include shell2junit library
+source <(curl -fsS --retry 3 'https://raw.githubusercontent.com/kubernetes/kubernetes/master/third_party/forked/shell2junit/sh2ju.sh')
+
+# Have cmd/e2e run by goe2e.sh generate JUnit report in ${WORKSPACE}/junit*.xml
+ARTIFACTS=${WORKSPACE}/_artifacts
+mkdir -p ${ARTIFACTS}
+
+# E2E runner stages
+STAGE_PRE="PRE-SETUP"
+STAGE_SETUP="SETUP"
+STAGE_CLEANUP="CLEANUP"
+STAGE_KUBEMARK="KUBEMARK"
+
 : ${KUBE_GCS_RELEASE_BUCKET:="kubernetes-release"}
 
+# record_command runs the command and records its output/error messages in junit format
+# it expects the first argument to be the class and the second to be the name of the command
+# Example:
+# record_command PRESETUP curltest curl google.com
+# record_command CLEANUP check false
+#
+# WARNING: Variable changes in the command will NOT be effective after record_command returns.
+#          This is because the command runs in subshell.
+function record_command() {
+    set +o xtrace
+    set +o nounset
+    set +o errexit
+
+    local class=$1
+    shift
+    local name=$1
+    shift
+    echo "Recording: ${class} ${name}"
+    echo "Running command: $@"
+    local -r ret=$(juLog -output="${ARTIFACTS}" -class="${class}" -name="${name}" "$@")
+
+    set -o nounset
+    set -o errexit
+    set -o xtrace
+
+    # since record_command disable errexit, exit script on failure.
+    local -r found=$(echo "${ret}" | grep -c "+++ exit code: 0")
+    if [[ ${found} -le 0 ]]; then
+        echo "Last command failed, exiting..."
+        exit 1
```

Comment:
Hold on that. @fejta is trying to convince us that we shouldn't care about
the difference and that it's perfectly valid to fail like this.

On Tue, Jun 28, 2016 at 3:23 PM, Daniel Smith <notifications@github.com>
wrote:

> In hack/jenkins/e2e-runner.sh
> <https://github.com/kubernetes/kubernetes/pull/28133#discussion_r68855336>
> :
>
> > +    shift
> > +    local name=$1
> > +    shift
> > +    echo "Recording: ${class} ${name}"
> > +    echo "Running command: $@"
> > +    local -r ret=$(juLog -output="${ARTIFACTS}" -class="${class}" -name="${name}" "$@")
> > +
> > +    set -o nounset
> > +    set -o errexit
> > +    set -o xtrace
> > +
> > +    # since record_command disable errexit, exit script on failure.
> > +    local -r found=$(echo "${ret}" | grep -c "+++ exit code: 0")
> > +    if [[ ${found} -le 0 ]]; then
> > +        echo "Last command failed, exiting..."
> > +        exit 1
>
> The goal here is to get a build marked "unstable" instead of "failed"
> which I understand means, now that I unfortunately learned about jenkins,
> that we shouldn't exit at all (we want to run all the commands) and should
> print "Test Suite Failed" so that jenkins will mark it unstable (@ixdy
> <https://github.com/ixdy> is about to turn that plugin on). @ixdy
> <https://github.com/ixdy>, sound correct?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/28133/files/498735ea3364f419c967944e9188dba4268b6ca4#r68855336>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAnglgA-mudyhnqfiFTu9cPQ0RlIcgr5ks5qQZ7_gaJpZM4I_jtj>
> .
>